### ISSUE TITLE
SELinux investigation & support

### ISSUE THREAD

**Issue by seungsoo-lee** [kubearmor/KubeArmor#76]
AWS EKS does not support AppArmor. They only support SELinux.

So, we need to investigate SELinux and support it in KubeArmor.

Update (5th Nov 2024): AWS EKS has since past couple of years started supporting BPF LSM out of the box and hence AWS EKS problem is no more an issue with SELinux.

**Comment by nam-jaehyun** [comment]
In terms of SELinux integration, we have limited features for containers due to SELinux labeling issues.
Thus, while dealing with the SELinux limitations, we want to provide policy-based auditing using eBPF.

For this, we need to handle two things.
1. Block policy -> generate alerts with "Block (Audit)" action
2. Allow policy -> generate alerts against all things except for allowed ones with "Allow (Audit)" action

**Comment by nam-jaehyun** [comment]
1. Observations

- KubeArmor Image
  The current image of KubeArmor is based on Alpine, and it doesn't provide SELinux-related binaries (e.g., semanage).
  Thus, to use 'semanage', we need to use some base images (e.g., centos:7).
  Note that if the image is based on CentOS, it cannot provide AppArmor-related binaries.

- SELinux status
  Even though we have an image based on CentOS, KubeArmor cannot control SELinux inside of the image.
  ```
  # sestatus
  SELinux status:                 disabled
  ```
  ```
  # semange /KubeArmor/templates/base_container.cil
  ValueError: SELinux policy is not managed or store cannot be accessed.
  ```
  Note that we already have the full privilege in the CentOS-based KubeArmor container.
  ```
  hostPID: true
  hostNetwork: true
  containers:
  - name: kubearmor
    securityContext:
      privileged: true
  ```

2. Conclusion
  We can support Redhat-oriented OSes (e.g., CentOS, Fedora, and RHEL) with KueArmor's audit mode.
  In other words, if we use the KubeArmor image, KubeArmor will automatically use the audit mode in such environments
  while it will provide security enforcement features in some other environments (e.g., Debian, Ubuntu, and Google COS).
  If we want to use SELinux features (e.g., access control against mounted host volumes),
  we need to run KubeArmor directly on top of Redhat-oriented OSes. Then, KubeArmor can support the features.

**Comment by nyrahul** [comment]
[Document prepared](https://docs.google.com/document/d/1tXg4LhVBS74fxafDt1vKmYakd3AzF9c405haq-yWtbc/edit) by @oneiro-naut for SELinux deployment in Kubearmor and limitations. Keeping here for ref.

**Comment by nyrahul** [comment]
[SELinux Test Analysis Design.docx](https://github.com/kubearmor/KubeArmor/files/8237664/SELinux.Test.Analysis.Design.docx)

These scenarios were tested by @ErenYeager713 . Thanks

**Comment by nam-jaehyun** [comment]
Reference: https://github.com/nam-jaehyun/KubeArmor/blob/main/examples/vagrant-selinux.md

**Comment by nam-jaehyun** [comment]
Fix: #673

**Comment by nyrahul** [comment]
Just for the record, [attaching herewith a document](https://docs.google.com/document/d/1AWVvScW8E1FiQPwJYia_ZnhdM_V-68dyjKwxhKuDZTM/edit?tab=t.0) that explains why using SELinux for granular enforcement policies is technically a very complex problem.

SELinux has something called as type enforcement i.e., you set a label on the resource and type enforcement uses this label for enforcement perspective. A change in the type enforcement rule has a trickle effect wherein it is difficult to judge what rules might be impacted. The attached document explains few other problems in the context of use of SELinux for granular policy enforcement.

[SELinux support in KubeArmor.pdf](https://github.com/user-attachments/files/17628751/SELinux.support.in.KubeArmor.pdf)
